pkg/cmd/cli/cmd/set/env.go

```diff
@@ -139,6 +151,68 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
 		return err
 	}
 
+	var sources []string
+	if secretName != "" {
+		sources = append(sources, fmt.Sprintf("secret/%s", secretName))
+	}
+	if configMapName != "" {
+		sources = append(sources, fmt.Sprintf("configmap/%s", configMapName))
+	}
+
+	if len(sources) != 0 {
+		mapper, typer := f.Object(false)
+		b := resource.NewBuilder(mapper, typer, resource.ClientMapperFunc(f.ClientForMapping), kapi.Codecs.UniversalDecoder()).
+			ContinueOnError().
+			NamespaceParam(cmdNamespace).DefaultNamespace().
+			FilenameParam(explicit, false, filenames...).
+			SelectorParam(selector).
+			ResourceTypeOrNameArgs(all, sources...).
+			Flatten()
+
+		one := false
+		infos, err := b.Do().IntoSingular(&one).Infos()
```

Comment:
hrmm, I see it is required to specify one env per key... so it snapshots the env to the keys that exist at this point in time, and requires access to the secret or configmap via the API